### DISCUSSION TITLE
Calling `taichi` kernels via primitives over `custom_call`

### DISCUSSION THREAD

**Discussion by femtomc** [jax-ml/jax#12162] [Q&A] [answered]
Has anyone investigated a route for interop between [`taichi`](https://github.com/taichi-dev/taichi) and JAX by utilizing `custom_call` infrastructure in XLA?

Are there immediate blockers to attempting this?

**Comment by sharadmv** [comment, marked as answer]
I actually did investigate adding Taichi custom calls but didn't made much progress and have not worked on it for a month or so. I am still struggling to understand how/where in their codebase Taichi launches the kernels. If you happen to know this, please let me know!

If you wanted to investigate this yourself, I think looking at [`jax-triton`](https://github.com/jax-ml/jax-triton) is a good starting point since it is a similar type of integration.

**Reply by femtomc** [reply]
Thank you @sharadmv -- good to know this is somewhat novel ground :)

**Comment by chaoming0625** [comment]
Dear @sharadmv , are you still working on Taichi? I think Taichi is much more flexible than triton when working on sparse data.

**Reply by sharadmv** [reply]
Unfortunately, no, I haven't time yet, though I agree that a Taichi integration would be really neat and covers a much different use-case from  Triton. Contributions are welcome and I'm happy to advise!

**Reply by chaoming0625** [reply]
What's the core problem which prevents us from integrating a Taichi kernel into the JAX JIT system? 

`jax-triton` utilizes the kernel information (`kernel_ptr`) to launch the function. Is this kernel pointer the key for Taichi integration?

**Reply by chaoming0625** [reply]
Currently, Taichi can be used to write a PyTorch operator. See example in [here](https://github.com/ailzhang/blog_code/blob/master/rwkv/benchmark.py#L45-L73). I think it is urgent to support the Taichi compiled kernels in JAX.

**Reply by femtomc** [reply]
@chaoming0625 I think the question is whether Taichi exposes a `kernel_ptr` that you could wrap in a `custom_call` invocation (if I'm understanding above).

In general, I think we want to wrap the kernel launch invocation in custom call -- we also need to provide shape invariants to that custom call, etc.

**Reply by chaoming0625** [reply]
Yes, I agree.

**Reply by femtomc** [reply]
I don't know enough about Taichi to know where to look for such a `kernel_ptr`, however :)

**Reply by chaoming0625** [reply]
I opened an issue in [Taichi repo](https://github.com/taichi-dev/taichi/issues/6367).  Hoping Taichi developers can give us some insights.

**Reply by sharadmv** [reply]
Yes my initial struggle with the Taichi was in obtaining the kernel pointer, which seems to be hidden away behind some abstractions. Thank you for following up!
